### PR TITLE
Support distributed evaluation

### PR DESCRIPTION
**Pull request by mehdidc**
Currently, evaluation is done on rank zero. This PR provides support for distributed evaluation (using an optional `--distributed-evaluation` argument) to make evaluation faster (supports both zero-shot and retrieval).

### CODE REVIEW

**Comment by rom1504** [comment]
this LGTM

could you please double check @mitchellnw or @rwightman ?

**Review by rom1504 — APPROVED** [review]

**Comment by mitchellnw** [comment]
In the pytorch imagenet example for distributed imagenet eval they have an `aux_val_loader` to handle the case where the test set size is not divisible by num_gpus - do we need to have this? https://github.com/pytorch/examples/blob/main/imagenet/main.py#L396-L402

File: src/training/train.py [hunk 1]
```diff
@@ -232,22 +255,24 @@ def evaluate(model, data, epoch, args, tb_writer=None):
...
+    if args.wandb and is_master(args):
         assert wandb is not None, 'Please install wandb.'
         for name, val in metrics.items():
             wandb.log({f"val/{name}": val, 'epoch': epoch})
 
     return metrics
 
 
-def get_metrics(image_features, text_features, logit_scale):
+def get_metrics(image_features, text_features, logit_scale, args):
```
Comment (usuyama) [inline src/training/train.py:266]: nit pick but maybe better to pass a more specific argument e.g. `gather_tensor=args.distributed_evaluation`?

File: src/training/data.py [hunk 2]
```diff
@@ -359,7 +361,8 @@ def get_wds_dataset(args, preprocess_img, is_train, epoch=0, floor=False):
         num_samples = num_batches * global_batch_size
         dataset = dataset.with_epoch(num_worker_batches)  # each worker is iterating over this
     else:
-        # last batches are partial, eval is done on single (master) node
+        if args.distributed_evaluation:
+            num_samples = num_samples // args.world_size
```
Comment (usuyama) [inline src/training/data.py:365]: maybe I'm misunderstanding, but does it skip the last few samples due to the last partial batch?

File: src/training/train.py (hunk 1 above)
Comment (mehdidc) [reply src/training/train.py:266]: > In the pytorch imagenet example for distributed imagenet eval they have an `aux_val_loader` to handle the case where the test set size is not divisible by num_gpus - do we need to have this? https://github.com/pytorch/examples/blob/main/imagenet/main.py#L396-L402

Thanks for the link, not sure why they do `drop_last=True` on val_loader (not used here), probably to avoid having a GPU worker with much fewer examples than the others? so rather, they seem to do drop_last, and compute the last few examples validation performance in all GPU workers.

Comment (mehdidc) [reply src/training/train.py:266]: > nit pick but maybe better to pass a more specific argument e.g. `gather_tensor=args.distributed_evaluation`?

Thanks I agree, I think it's better

File: src/training/data.py (hunk 2 above)
Comment (mehdidc) [reply src/training/data.py:365]: > maybe I'm misunderstanding, but does it skip the last few samples due to the last partial batch?

in evaluation, `num_samples` is only used on logging (https://github.com/mlfoundations/open_clip/blob/03839c5e00b1c21c14dbaa07921d7190a28125c5/src/training/train.py#L172, https://github.com/mlfoundations/open_clip/blob/03839c5e00b1c21c14dbaa07921d7190a28125c5/src/training/train.py#L217), it is not affecting the dataloader which depend on the wds pipeline. But it would be good to get it correct anyway, have to check how many examples  each worker exactly receive.

**Comment by mehdidc** [comment]
> In the pytorch imagenet example for distributed imagenet eval they have an `aux_val_loader` to handle the case where the test set size is not divisible by num_gpus - do we need to have this? https://github.com/pytorch/examples/blob/main/imagenet/main.py#L396-L402

Thanks for the link, not sure why they do drop_last=True on val_loader (not used here), probably to avoid having a GPU worker with much fewer examples than the others? so rather, they seem to do drop_last, and compute the last few examples validation performance in all GPU workers.

**Comment by mitchellnw** [comment]
@mehdidc I think this is actually necessary or else you can get different val perf when different numbers of gpus are used, e.g., see this comment: https://github.com/facebookresearch/deit/blob/main/main.py#L221-L223

**Comment by mehdidc** [comment]
> @mehdidc I think this is actually necessary or else you can get different val perf when different numbers of gpus are used, e.g., see this comment: https://github.com/facebookresearch/deit/blob/main/main.py#L221-L223

I see, thanks @mitchellnw! OK so I need to fix this. I really thought that `DistributedSampler` with `drop_last=False` would do the the "right" thing (although wouldn't be ideal for disributed setting if it would be case) in the sense of seing each example exactly once (even when dataset size is not divisible by nb of workers, e.g. the remaining examples can be distributed to a subset of workers)

**Comment by mehdidc** [comment]
@dmlpt Not yet, I still need to fix the val dataloader like @mitchellnw mentions
